[API] Finalize the first draft of C-API

kparichay opened the pull request:
Finalize the first draft of C-API

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

zhoonit commented on api/capi/include/nntrainer.h:
```diff
@@ -65,9 +72,31 @@ typedef enum {
+ * @retval #ML_ERROR_NONE Successful.
+ * @retval #ML_ERROR_INVALID_PARAMETER Invalid Parameter.
+ */
+typedef int (*dataset_generator) (float *input, float *label, bool *last);
```
I think there should be a doc about format of the data.

sth like:
1. It should match the shape of specified input
2. data is read row-wise

kparichay replied:
1. I am thinking of adding a `size_t bytes` in this function for this, as this function should know about the size of the allocated memory. Yes, added a comment for shape.
2. Added.

kparichay replied:
Updated `float *input`  and `float *label` to `float **input` and `float **label` to support multiple inputs and outputs.

zhoonit commented on api/capi/include/nntrainer.h:
```diff
@@ -65,9 +72,31 @@ typedef enum {
+ * @note This function can be called multiple times in parallel.
+ * @param[out] input Container to provide input data to the model.
+ * @param[out] label Container to provide label data to the model to train/test.
+ * @param[out] last Container to provide is the data is finished.
```
I feel hard to catch what this does from the description

kparichay replied:
Updated description.

zhoonit replied:
Nice :) Just one minor thing.
```suggestion
 * @param[out] last Container to provide is the data finished.
```

zhoonit commented on api/capi/include/nntrainer.h:
```diff
@@ -65,9 +72,31 @@ typedef enum {
+ * @param[in] cb The callback to be called once the training is finished.
+ * @param[in] data Internal data to be given to the callback, cb.
+ */
+typedef void (*train_notify) (ml_nnmodel_h model, void *data);
```
AFAIK, train runs synchronously for now. Could you check this out?

kparichay replied:
Yes, it will still be synchronous. Their will be wrapper around it to make it asynchronous. This for `ml_nnmodel_train_async`

zhoonit replied:
Sorry for that. I didn't read that part yet. Thanks for the quick response :)

zhoonit commented on api/capi/include/nntrainer.h:
```diff
@@ -89,8 +118,12 @@ int ml_nnmodel_construct(ml_nnmodel_h *model);
- * @details Use this function to initialize neural network model
+ * @brief Compile and finalize the neural network model.
+ * @details Use this function to initialize neural network model. Once compiled,
+ * updating the model will be restricted. Any modification to the
```
```suggestion
 * any modification to the
```

kparichay replied:
Done.

zhoonit commented on api/capi/include/nntrainer.h:
```diff
@@ -227,6 +289,64 @@ int ml_nnoptimizer_delete(ml_nnopt_h opt);
+ * @retval #ML_ERROR_NONE Successful.
+ * @retval #ML_ERROR_INVALID_PARAMETER Invalid Parameter.
+ */
+int ml_nndataset_create_with_generator(ml_nndataset_h *dataset,
```
Would there be usecases that user doing only testing without training?

kparichay replied:
With the current setup, `train_cb` should be necessarily provided. So, not allowed for now.
Later, can support train_cb to be null to support testing without training.

zhoonit replied:
Reasonable.

zhoonit commented on api/capi/include/nntrainer.h:
```diff
@@ -65,9 +72,31 @@ typedef enum {
+ * space to contain the data by the caller. This function should return a batch
+ * of data in the passed containers.
+ * @note This function can be called multiple times in parallel.
+ * @param[out] input Container to provide input data to the model.
```
@param[in] for those?

jijoongmoon commented on api/capi/include/nntrainer.h:
```diff
@@ -65,9 +72,33 @@ typedef enum {
+ * @param[out] input Container to hold all the input data.
+ * @param[out] label Container to hold corresponding label data.
+ * @param[out] last Container if data is finished. Set true if this data is the last piece, else set false.
+ * @param[in] batch Number of batches of data needed.
```
no param name with batch. remove?

kparichay replied:
Removed.

jijoongmoon commented on api/capi/include/nntrainer.h:
```diff
@@ -65,9 +72,33 @@ typedef enum {
+
+/**
+ * @brief Callback function to notify completion of training of the model..
+ * @param[in] cb The callback to be called once the training is finished.
```
need to change? no param[in] with cb

kparichay replied:
Updated.

jijoongmoon commented on api/capi/include/nntrainer.h:
```diff
@@ -227,6 +290,64 @@ int ml_nnoptimizer_delete(ml_nnopt_h opt);
+ * files.
+ * @since_tizen 6.x
+ * @param[out] dataset The NNTrainer Dataset handler from the given description.
+ * @param[in] train_cb The dataset file for training.
```
this is not cb. train_file? same with valid_file and test_file.

kparichay replied:
Updated.

jijoongmoon commented on api/capi/include/nntrainer.h:
```diff
@@ -227,6 +290,64 @@ int ml_nnoptimizer_delete(ml_nnopt_h opt);
+
+/**
+ * @brief Set the neural network dataset Property.
+ * @details Use this function to set dataset Property.
```
what kind of data set properties are? is it something like buffer_size, mini_batch size, etc.?

kparichay replied:
Yes - buffer size, num data loader threads, shuffle, etc. 
This function allows to extend the properties later if needed.

kparichay commented:
**V2:**
Applied review comments
Updated namespace to `ml_train_*`
Added module doc
Added enums for loss and optimizer
Added loss as a parameter for model_compile

myungjoo commented on api/capi/doc/nntrainer_doc.h:
```diff
@@ -0,0 +1,68 @@
+
+/**
+ * @ingroup  CAPI_ML_FRAMEWORK
+ * @defgroup CAPI_ML_NNTRAINER_TRAIN_MODULE Pipeline
```
```s/@defgroup CAPI_ML_NNTRAINER_TRAIN_MODULE Pipeline/@defgroup CAPI_ML_NNTRAINER_TRAIN_MODULE Trainer/```?

myungjoo replied:
@kparichay @jijoongmoon @zhoonit 
You need to declare group name. e.g., Trainer, NNTrainer, OnDeviceTrainer, ... ... You shouldn't use Pipeline as the name of this group.

kparichay replied:
Updated name of the group to Trainer.

myungjoo commented on api/capi/doc/nntrainer_doc.h:
```diff
@@ -0,0 +1,68 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
SPDX is recommended. You don't need the full license declaration.

kparichay replied:
Updated.

kparichay commented:
> We need to declare somewhere (or everywhere) that this API set is experimental.

Added to api/capi/include/nntrainer.h and api/capi/include/nntrainer_internal.h

myungjoo commented on api/capi/include/nntrainer.h:
```diff
@@ -13,12 +13,16 @@
+ * @date 08 July 2020
  * @brief NNTrainer C-API Header.
- *        This allows to construct and control NNTrainer Model.
+ *        This allows to construct, control and train NNTrainer Model.
```
```
This allows to construct, control and train NNTrainer Model.
```
===>
```
This allows to construct, control, and train a neural network model in Tizen devices.
```

Let's not restrict model format here.

kparichay replied:
done.
